### PR TITLE
Make json handle all types

### PR DESCRIPTION
**Pull request by breedx-nr**
This primarily addresses the concern that the `TelemetryBatchJson` class violated its contract and could only handle metrics.  This fixes that situation and the `TelemetryBatchJson` is no longer coupled to Metrics.

### CODE REVIEW

File: metrics/src/main/java/com/newrelic/telemetry/json/TelemetryBatchJson.java [hunk 1]
```diff
@@ -12,15 +12,16 @@
...
 
-/** Class that convert any type of telemetry batch into the appropriate API json */
+/** Class that converts any type of telemetry batch into the appropriate API json */
 public class TelemetryBatchJson {
 
   private static final Logger logger = LoggerFactory.getLogger(TelemetryBatchJson.class);
-  private final JsonCommonBlockWriter commonBlockWriter;
-  private final JsonTelemetryBlockWriter mainBodyWriter;
+  private final TypeDispatchingJsonCommonBlockWriter commonBlockWriter;
+  private final TypeDispatchingJsonTelemetryBlockWriter mainBodyWriter;
```
Comment (breedx-nr) [inline metrics/src/main/java/com/newrelic/telemetry/json/TelemetryBatchJson.java]: rename to `telemtryBlockWriter` for consistency.

File: metrics/src/main/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriter.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  private <T extends Telemetry> JsonTelemetryBlockWriter chooseMainBodyWrite(
+      TelemetryBatch<T> batch) {
+    switch (batch.getType()) {
+      case METRIC:
+        return mainBodyMetricsWriter;
+      case SPAN:
+        return mainBodySpanWriter;
+    }
+    throw new UnsupportedOperationException("Unhandled metric batch type: " + batch.getType());
```
Comment (breedx-nr) [inline metrics/src/main/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriter.java]: Phrasing is inconsistent (should be `telemetry`)

File: metrics/src/main/java/com/newrelic/telemetry/json/MetricBatchJsonTelemetryBlockWriter.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+public class MetricBatchJsonTelemetryBlockWriter implements JsonTelemetryBlockWriter {
+
+  private final MetricToJson metricToJson;
+
+  public MetricBatchJsonTelemetryBlockWriter(MetricToJson metricToJson) {
+    this.metricToJson = metricToJson;
+  }
+
+  @Override
+  public <T extends Telemetry> void appendTelemetry(TelemetryBatch<T> batch,
```
Comment (jeemar) [inline metrics/src/main/java/com/newrelic/telemetry/json/MetricBatchJsonTelemetryBlockWriter.java]: Maybe `appendTelemetryJson` to match the other interface method

File: metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonCommonBlockWriterTest.java [hunk 4]
```diff
@@ -34,7 +29,18 @@ void testMetrics() {
 
   @Test
   void testSpans() {
-    fail("build me");
+    StringBuilder sb = new StringBuilder();
+    TelemetryBatch<Span> batch = mock(TelemetryBatch.class);
+    JsonCommonBlockWriter spansWriter = mock(JsonCommonBlockWriter.class);
+
+    when(batch.getType()).thenReturn(Type.SPAN);
```
Comment (jeemar) [inline metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonCommonBlockWriterTest.java]: Add test for the unsupported type

File: metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriterTest.java [hunk 5]
```diff
@@ -1,19 +1,45 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
+import com.newrelic.telemetry.Metric;
+import com.newrelic.telemetry.Span;
+import com.newrelic.telemetry.Telemetry.Type;
+import com.newrelic.telemetry.TelemetryBatch;
 import org.junit.jupiter.api.Test;
 
 class TypeDispatchingJsonTelemetryBlockWriterTest {
```
Comment (jeemar) [inline metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriterTest.java]: Add test for the unsupported `batch.getType()`

File: metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonCommonBlockWriterTest.java [hunk 6]
```diff
@@ -1,30 +1,25 @@
...
-import com.newrelic.telemetry.Gauge;
 import com.newrelic.telemetry.Metric;
+import com.newrelic.telemetry.Span;
 import com.newrelic.telemetry.Telemetry.Type;
 import com.newrelic.telemetry.TelemetryBatch;
-import java.util.Collection;
-import java.util.Collections;
 import org.junit.jupiter.api.Test;
 
 class TypeDispatchingJsonCommonBlockWriterTest {
```
Comment (jeemar) [inline metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonCommonBlockWriterTest.java]: Add test for the unsupported batch.getType()

File: metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriterTest.java (hunk 5 above)
Comment (breedx-nr) [reply metrics/src/test/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriterTest.java]: Due to the nature of the enum, this is very challenging and I don't think it's worth doing.  If you have an interesting way of getting coverage, I'd love to hear it!

File: metrics/src/main/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriter.java (hunk 2 above)
Comment (breedx-nr) [reply metrics/src/main/java/com/newrelic/telemetry/json/TypeDispatchingJsonTelemetryBlockWriter.java]: Fixed.

File: metrics/src/main/java/com/newrelic/telemetry/json/MetricBatchJsonTelemetryBlockWriter.java (hunk 3 above)
Comment (breedx-nr) [reply metrics/src/main/java/com/newrelic/telemetry/json/MetricBatchJsonTelemetryBlockWriter.java]: Good idea, fixed that.  Thanks!

File: metrics/src/main/java/com/newrelic/telemetry/json/TelemetryBatchJson.java (hunk 1 above)
Comment (breedx-nr) [reply metrics/src/main/java/com/newrelic/telemetry/json/TelemetryBatchJson.java]: Fixed.

**Pull request merged**